crypto/sm2/sm2_crypt.c

```diff
@@ -114,26 +115,31 @@ int sm2_encrypt(const EC_KEY *key,
     EVP_MD_CTX *hash = EVP_MD_CTX_new();
     struct SM2_Ciphertext_st ctext_struct;
     const EC_GROUP *group = EC_KEY_get0_group(key);
-    const BIGNUM *order = EC_GROUP_get0_order(group);
+    const BIGNUM *order;
     const EC_POINT *P = EC_KEY_get0_public_key(key);
     EC_POINT *kG = NULL;
     EC_POINT *kP = NULL;
     uint8_t *msg_mask = NULL;
     uint8_t *x2y2 = NULL;
     uint8_t *C3 = NULL;
     const size_t field_size = ec_field_size(group);
-    const size_t C3_size = EVP_MD_size(digest);
+    const int C3_size = EVP_MD_size(digest);
 
     /* NULL these before any "goto done" */
     ctext_struct.C2 = NULL;
     ctext_struct.C3 = NULL;
 
     if (hash == NULL
             || group == NULL
-            || order == NULL
```

Comment:
Side note. Formally speaking one can write (order = EC_GROUP_get0_order(group)) == NULL right here. Because if group is NULL, then evaluation is guaranteed to stop. Having them next to each other might be considered as more readable.

Reply:
On related thought one can wonder if all of these checks are actually justified. In sense that can one actually instantiate EC_KEY or EC_GROUP without all/some of those members? In other words can one actually instantiate EC_KEY without group? Can there be EC_GROUP without order? Point is that if one can't do that, then checks would be excessive/redundant. As they are performed by the time EC_KEY is instantiated.

Reply:
I would say they are justified. It is possible to create an `EC_KEY` without those members. Creating an EC_GROUP without an order is less clear as it depends on the behaviour of flags and method initialisation functions.

Reply:
While it's indeed [apparently] possible to create EC_KEY object without group, question also is if application should be permitted to pass such key *here*. I mean just like memcpy doesn't check for pointers being non-NULL, it's not unreasonable to demand that EC_KEY passed *here* does have group. In the context one can even say that error message would be misleading, because it reads "*internal* error", while the problem is actually caller's. In other words I'm still not convinced that checks are not redundant, but whatever...

Here is another problem. Call to ec_field_size few lines up. If one is so picky, then it should be moved past NULL check as well. Common problem is that one tends to think about how to appease coverity, while one should concentrate on what code does or should do.

On related note one can actually wonder if ec_field_size implementation is right. I mean it calls EC_GROUP_get_curve_GFp, i.e. makes quite specific **assumption** about curve type [right?]...

Reply:
I moved the call to `ec_field_size` past the NULL check.

WRT assumption about the curve type, it's not clear to me whether this is correct (and SM2 is restricted to GFp), or whether this is an error. I created an issue (#6646) for it.